### PR TITLE
Use internal API server service (Without ClusterIP) + Do custom endpoint reconciling to avoid handling the NodePort

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/cluster/userclusterresources.go [hunk 1]
```diff
@@ -601,3 +601,68 @@ func (cc *Controller) userClusterEnsureServices(c *kubermaticv1.Cluster) error {
 
 	return nil
 }
+
+func (cc *Controller) reconcileAPIServerEndpoints(c *kubermaticv1.Cluster) error {
```
Comment (alvaroaleman) [inline api/pkg/controller/cluster/userclusterresources.go:605]: IMHo this should run inside the user cluster controller manages as a dedicated controller and also have a lister for the services in the seed NS of the cluster to ensure we react quickly to any changes

File: api/pkg/resources/apiserver/service.go [hunk 2]
```diff
@@ -20,15 +20,16 @@ func Service(data resources.ServiceDataProvider, existing *corev1.Service) (*cor
 	se.Labels = resources.BaseAppLabel(name, nil)
 
 	se.Spec.Type = corev1.ServiceTypeClusterIP
+	se.Spec.ClusterIP = corev1.ClusterIPNone
```
Comment (alvaroaleman) [inline api/pkg/resources/apiserver/service.go:23]: Why?

**Review by alvaroaleman — COMMENTED** [review]
Would love to see the `Use internal API server service` part.

Is there a reason why that must be done together with the custom endpoint reconciling?

**Comment by mrIncompetent** [comment]
This PR does change 3 things (Which i'm not certain if they are worth it):
- Use the `apiserver` service instead of the `apiserver-external` service
  - That itself should not have any impact. The `apiserver-external` & `apiserver` are both ClusterIP services after all.
- Use RoundRobin DNS (`clusterIP=none`) instead of a ClusterIP for the `apiserver` service. This was mostly a "esoteric" change with the hope of better connection handling for clients.
- Use custom endpoint reconciliation 
  - This is required since we now use RoundRobin DNS instead of a ClusterIP. As we cannot re-map ports of a service.

Just using the `apiserver` service instead of the `apiserver-external` has technically no impact at all.

We might close this PR after all as it seems our connection issues might have been caused by GKE.

**Comment by mrIncompetent** [comment]
@alvaroaleman 
Please give some feedback on https://github.com/kubermatic/kubermatic/pull/2690#issuecomment-463147474

Otherwise i'll close this PR by the end of the week

**Comment by alvaroaleman** [comment]
> This PR does change 3 things (Which i'm not certain if they are worth it):

> Use the apiserver service instead of the apiserver-external service
That itself should not have any impact. The apiserver-external & apiserver are both ClusterIP services after all.
Use RoundRobin DNS (clusterIP=none) instead of a ClusterIP for the apiserver service. This was mostly a "esoteric" change with the hope of better connection handling for clients.
Use custom endpoint reconciliation
This is required since we now use RoundRobin DNS instead of a ClusterIP. As we cannot re-map ports of a service.
Just using the apiserver service instead of the apiserver-external has technically no impact at all.

> We might close this PR after all as it seems our connection issues might have been caused by GKE.

The reason I wanted this is that I do not want our controller manager to do the extra hop out of the seed and back in when it talks to user clusters. But this PR does not achieve that anyways so I'd vote for just closing it. If you do want it, just re-open.

**Pull request closed without merging**